x/gamm/pool-models/balancer/amm.go

```diff
@@ -250,16 +250,25 @@ func (p *Pool) calcSingleAssetJoin(tokenIn sdk.Coin, swapFee sdk.Dec, tokenInPoo
 	).TruncateInt(), nil
 }
 
+// JoinPool calculates the number of shares needed and the updated liquidity for joining pool
+// and updates pool accordingly.
 func (p *Pool) JoinPool(_ctx sdk.Context, tokensIn sdk.Coins, swapFee sdk.Dec) (numShares sdk.Int, err error) {
 	numShares, newLiquidity, err := p.CalcJoinPoolShares(_ctx, tokensIn, swapFee)
 	if err != nil {
 		return sdk.Int{}, err
 	}
+
+	// update pool with the calculated share and liquidity needed to join pool
 	p.IncreaseLiquidity(numShares, newLiquidity)
 	return numShares, nil
 }
 
-// CalcJoinPoolShares
+// CalcJoinPoolShares calculates the number of shares created to join pool with the provided amount of `tokenIn`.
+// When a single token is provided as an argument, we simply perform single asset join with the token.
+// If tokenIn provided as an argument isn't a sinlge token, it must contain all the tokens in the pool.
+// For the case of multi-asset join for a pool, we first calculate the maximum amount we can join a pool without swap, then
+// perform single asset join for the remaining coins.
+// CalcJoinPoolShares does not directly alter the state of the pool, but only does the calculation for shares for joining the pool.
 func (p *Pool) CalcJoinPoolShares(_ sdk.Context, tokensIn sdk.Coins, swapFee sdk.Dec) (numShares sdk.Int, newLiquidity sdk.Coins, err error) {
```

Comment:
This is not specific to this PR. I just kept thinking about this while working on this method over these past few days. Basically, I think that we should have separate methods for single tokenIn calc and multi tokensIn calc.

The way it is done currently is violating the "Single responsibility principle" in SOLID where the method does 2 things.
It makes it less testable and less abstract IMO. We should consider addressing this post v10. Please create an issue if you agree. If not, please ignore

Reply:
Agree

Reply:
Agreed